### PR TITLE
[Feature] Support STARK training

### PR DESCRIPTION
**Pull request by JingweiZhang12**

### CODE REVIEW

**Comment by JingweiZhang12** [comment]
1. the commit [6dd98ad](https://github.com/open-mmlab/mmtracking/pull/434/commits/6dd98ad30b4d3f587f6c3058c5b45f664018d616) is about the sampler.
2. The head of STARK doesn't have functions such as `get_targets` like siamprn++,  and the unit test of stark head is all included in the `test_stark_forward`, so I don't write repeatedly unit test about stark_head.py.

File: mmtrack/models/sot/stark.py [hunk 1]
```diff
@@ -0,0 +1,247 @@
...
+        self.backbone = build_backbone(backbone)
+        self.neck = build_neck(neck)
+        self.head = build_head(head)
+
+        self.test_cfg = test_cfg
+        self.train_cfg = train_cfg
+
+        # Set the update interval
+        self.update_intervals = self.test_cfg['update_intervals']
+        self.num_extra_template = len(self.update_intervals)
```
Comment (GT9505) [inline mmtrack/models/sot/stark.py:55]: `self.update_intervals` is only used in here.
unnecessary to record it

File: mmtrack/models/sot/stark.py [hunk 2]
```diff
@@ -0,0 +1,247 @@
...
+
+        self.debug = False
```
Comment (GT9505) [inline mmtrack/models/sot/stark.py:57]: delete this

File: mmtrack/models/sot/stark.py [hunk 3]
```diff
@@ -0,0 +1,247 @@
...
+        if frozen_modules is not None:
+            self.freeze_module(frozen_modules)
+        # We only train classification head in stage-2. The transformer and
+        # bbox head are not trained in stage-2.
+        if self.head.run_cls_head and not self.head.run_bbox_head:
+            for name, module in self.head.named_parameters():
```
Comment (GT9505) [inline mmtrack/models/sot/stark.py:63]: use `self.freeze_module` to freeze modules.
And `self.head.run_cls_head` and `self.head.run_bbox_head` can be ommited.

File: mmtrack/models/sot/stark.py [hunk 4]
```diff
@@ -0,0 +1,247 @@
...
+        # We don't use the `init_weights()` function in BaseModule, since it
+        # doesn't support the initialization method from `reset_parameters()`
+        # in Pytorch.
+        if self.with_backbone:
+            self.backbone.init_weights()
+
+        if self.with_neck:
+            for m in self.neck.modules():
+                if isinstance(m, _ConvNd) or isinstance(m, _BatchNorm):
+                    m.reset_parameters()
```
Comment (GT9505) [inline mmtrack/models/sot/stark.py:78]: Are you sure that the `neck` need to use `reset_parameters` to init ?

File: mmtrack/models/sot/stark.py [hunk 5]
```diff
@@ -0,0 +1,247 @@
...
+                H and W are both equal to that of search image.
+
+            search_gt_labels (Tensor, optional): the categories of bbox:
+                positive(1) or negative(0). It's of shape (N, 2). We just use
+                search_gt_labels[:, 1].
+
+        Returns:
+            dict[str, Tensor]: a dictionary of loss components.
+        """
+        z1_dict = self.forward_before_head(img[:, 0], padding_mask[:, 0])
```
Comment (GT9505) [inline mmtrack/models/sot/stark.py:211]:
```suggestion
        z1_dict = self.extrat_feats(img[:, 0], padding_mask[:, 0])
```

File: mmtrack/models/sot/stark.py [hunk 6]
```diff
@@ -0,0 +1,247 @@
...
+        track_results = self.head(head_dict_inputs)
+
+        losses = dict()
+        if self.head.run_bbox_head:
+            img_size = search_img[:, 0].shape[2]
+            tracking_bboxes = track_results['pred_bboxes'][:, 0] / img_size
+            search_gt_bboxes = (
+                torch.cat(search_gt_bboxes, dim=0).type(torch.float32)[:, 1:] /
+                img_size).clamp(0., 1.)
+            head_losses = self.head.reg_loss(tracking_bboxes, search_gt_bboxes)
```
Comment (GT9505) [inline mmtrack/models/sot/stark.py:234]: The loss of  head should be computed in head, not here

File: mmtrack/models/sot/stark.py [hunk 7]
```diff
@@ -0,0 +1,247 @@
...
+                - 'pos_embed': of the position embedding of shape
+                    (N, C, H // stride, W // stride)
+        """
+        feat = self.backbone(img)
+        feat = self.neck(feat)[0]
+
+        # official code uses default 'bilinear' interpolation.
+        mask = F.interpolate(
+            mask[None].float(), size=feat.shape[-2:]).to(torch.bool)[0]
+        pos_embed = self.head.positional_encoding(mask)
```
Comment (GT9505) [inline mmtrack/models/sot/stark.py:148]: the `pos_embed` should be created in `head`, not here

File: mmtrack/models/sot/stark.py [hunk 8]
```diff
@@ -0,0 +1,247 @@
...
+
+        Returns:
+            dict[str, Tensor]: a dictionary of loss components.
+        """
+        z1_dict = self.forward_before_head(img[:, 0], padding_mask[:, 0])
+        z2_dict = self.forward_before_head(img[:, 1], padding_mask[:, 1])
+        x_dict = self.forward_before_head(search_img[:, 0],
+                                          search_padding_mask[:, 0])
+        inputs = [z1_dict, z2_dict, x_dict]
+        head_dict_inputs = self._merge_template_search(inputs)
```
Comment (GT9505) [inline mmtrack/models/sot/stark.py:216]: This line should be move into head.

**Review by GT9505 — COMMENTED** [review]
In the forward_train of stark, we only need to 
1. extract the features of imgs, 
2. pass the feature into the head
3. get the training targets from head
4. compute loss
you can refer to siamrpn

**Review by GT9505 — COMMENTED** [review]
If there is anything that makes you confused, you can discuss it with me.

File: mmtrack/models/track_heads/stark_head.py [hunk 9]
```diff
@@ -0,0 +1,472 @@
...
+                 num_query=1,
+                 transformer=None,
+                 positional_encoding=dict(
+                     type='SinePositionalEncoding',
+                     num_feats=128,
+                     normalize=True),
+                 run_cls_head=False,
+                 run_bbox_head=True,
+                 bbox_head=None,
+                 cls_head=None,
```
Comment (GT9505) [inline mmtrack/models/track_heads/stark_head.py:292]: can we use whether `cls_head` is None or not to decide whether `run_cls_head` is True or False

File: mmtrack/models/sot/stark.py (hunk 1 above)
Comment (JingweiZhang12) [reply mmtrack/models/sot/stark.py:55]: `self.update_intervals` is used in the test stage

File: mmtrack/models/sot/stark.py (hunk 3 above)
Comment (JingweiZhang12) [reply mmtrack/models/sot/stark.py:63]: The `backbone` and `neck` in the configs can be freezed by `self.freeze_module`. The `head ` module contains 'transformer'', `bbox_head` and `cls_head`, we can not directly used `self.freeze_module` to only freeze the `transformer` and `bbox_head` except `cls_head`.

File: mmtrack/models/sot/stark.py (hunk 4 above)
Comment (JingweiZhang12) [reply mmtrack/models/sot/stark.py:78]: I've checked it. It uses the default initialization method in the pytorch.

File: mmtrack/models/sot/stark.py (hunk 3 above)
Comment (GT9505) [reply mmtrack/models/sot/stark.py:61]: We can add a `freeeze_module` in the head.
In summary, we need to delete `run_bbox_head` and `run_cls_head`, since it is hard code.
How to freeze the module in the model should be configed flexiblely in the config.

File: mmtrack/models/sot/stark.py [hunk 10]
```diff
@@ -0,0 +1,188 @@
...
+            mask (Tensor): mask of shape (N, H, W).
+
+        Returns:
+            dict(Tensor):
+                - 'feat': the multi-level feature map of shape
+                    (N, C, H // stride, W // stride).
+                - 'mask': the mask of shape (N, H // stride, W // stride).
+        """
+        feat = self.backbone(img)
+        feat = self.neck(feat)[0]
```
Comment (GT9505) [inline mmtrack/models/sot/stark.py:96]:
```suggestion
        feat = self.neck(feat)
```
Directly pass `feat` into the `head` as a tuple.
Let the head deal with multi-level feature maps or single-level feature map

File: mmtrack/models/sot/stark.py [hunk 11]
```diff
@@ -0,0 +1,188 @@
...
+
+        # official code uses default 'bilinear' interpolation.
+        mask = F.interpolate(
```
Comment (GT9505) [inline mmtrack/models/sot/stark.py:99]: The same as above. 
Directly pass mask into the head.
If the `feat` contains multi-level feature maps, the mask needs to be resized multiple times.

File: mmtrack/models/sot/stark.py [hunk 12]
```diff
@@ -0,0 +1,188 @@
...
+            search_gt_labels (Tensor, optional): the categories of bbox:
+                positive(1) or negative(0). It's of shape (N, 2). We just use
+                search_gt_labels[:, 1].
+
+        Returns:
+            dict[str, Tensor]: a dictionary of loss components.
+        """
+        z1_dict = self.extract_feats(img[:, 0], padding_mask[:, 0])
+        z2_dict = self.extract_feats(img[:, 1], padding_mask[:, 1])
+        x_dict = self.extract_feats(search_img[:, 0], search_padding_mask[:,
```
Comment (GT9505) [inline mmtrack/models/sot/stark.py:165]: Pass original `padding_mask` into head. Just like `positional_embeding`

File: mmtrack/models/sot/stark.py [hunk 13]
```diff
@@ -0,0 +1,188 @@
...
+                    [tl_x, tl_y, br_x, br_y] format.
+            - 'pred_logits': bboxes of (N, num_query, 1) shape.
+        Typically `num_query` is equal to 1.
+        '''
+        track_results = self.head(head_inputs)
+
+        losses = dict()
+        head_losses = self.head.loss(track_results, search_gt_bboxes,
+                                     search_gt_labels, search_img[:,
+                                                                  0].shape[2])
```
Comment (GT9505) [inline mmtrack/models/sot/stark.py:181]: Are you sure using `search_img[:, 0].shape[2])` (only `H`)

File: mmtrack/models/track_heads/stark_head.py [hunk 14]
```diff
@@ -0,0 +1,554 @@
...
+        self.num_query = num_query
+        self.query_embedding = nn.Embedding(self.num_query, self.embed_dims)
+
+        self.train_cfg = train_cfg
+        self.test_cfg = test_cfg
+        self.fp16_enabled = False
+
+    def init_weights(self):
+        """Parameters initialization."""
+        self.transformer.init_weights()
```
Comment (GT9505) [inline mmtrack/models/track_heads/stark_head.py:327]: This function is unnecessary

File: mmtrack/models/track_heads/stark_head.py [hunk 15]
```diff
@@ -0,0 +1,554 @@
...
+            dict:
+                - 'pred_bboxes': of shape (bs, num_query, 4), in
+                    [tl_x, tl_y, br_x, br_y] format
+                - 'pred_logit': of shape (bs, num_query, 1)
+        """
+        out_dict = {}
+        if self.cls_head is not None:
+            # forward the classification head
+            out_dict['pred_logits'] = self.cls_head(feat)[-1]
+        if self.run_bbox_head:
```
Comment (GT9505) [inline mmtrack/models/track_heads/stark_head.py:433]: Delete `run_bbox_head`
use `if self.reg_head is not None:`

File: mmtrack/models/track_heads/stark_head.py [hunk 16]
```diff
@@ -0,0 +1,554 @@
...
+        bbox_feat = bbox_feat.permute((0, 3, 2, 1)).contiguous()
+        bs, num_query, dim, _ = bbox_feat.size()
+        bbox_feat = bbox_feat.view(-1, dim, self.bbox_head.feat_size,
+                                   self.bbox_head.feat_size)
+        # run the corner prediction head
+        outputs_coord = self.bbox_head(bbox_feat)
+        outputs_coord = outputs_coord.view(bs, num_query, 4)
+        return outputs_coord
+
+    def forward_head(self, feat, enc_mem):
```
Comment (GT9505) [inline mmtrack/models/track_heads/stark_head.py:409]: merge this function into `forward`

File: mmtrack/models/track_heads/stark_head.py [hunk 17]
```diff
@@ -0,0 +1,554 @@
...
+            gt_labels (Tensor): ground truth labels,
+                with shape (bs, 1).
+        Returns:
+            dict: a dictionary of classification loss components.
+        """
+        losses = dict()
+        losses['loss_cls'] = self.loss_cls(pred_logits, gt_labels)
+        return losses
+
+    def loss(self,
```
Comment (GT9505) [inline mmtrack/models/track_heads/stark_head.py:514]: merge `cls_loss` and `reg_loss` into `loss`

**Review by GT9505 — COMMENTED** [review]
Generally speaking, the `model` can only call the general API of `head`, such as `head.init_weights`, `head.loss` and so on.
The `model` should not call the unique attribute of `head`, such as `head.run_cls_head`, `head.run_reg_head`, `head.cls_head`, `head.reg_head` and so on.

File: mmtrack/models/track_heads/stark_head.py [hunk 18]
```diff
@@ -0,0 +1,515 @@
...
+            gt_bboxes[:, 0:4:2] = gt_bboxes[:, 0:4:2] / float(img_size[1])
+            gt_bboxes[:, 1:4:2] = gt_bboxes[:, 1:4:2] / float(img_size[0])
+            gt_bboxes = gt_bboxes.clamp(0., 1.)
+
+            # regression IoU loss, default GIoU loss
+            if (pred_bboxes[:, :2] >= pred_bboxes[:, 2:]).any() or (
+                    gt_bboxes[:, :2] >= gt_bboxes[:, 2:]).any():
+                # the first several iterations of train may return invalid
+                # bbox coordinates.
+                losses['loss_iou'] = torch.tensor(0.0).to(pred_bboxes)
```
Comment (GT9505) [inline mmtrack/models/track_heads/stark_head.py:504]: suggest to use `(pred_bboxes-gt_bboxes).sum()*0.0` to fix the bug

File: tests/test_models/test_forward/test_sot_forward.py [hunk 19]
```diff
@@ -12,7 +12,7 @@
     'sot/siamese_rpn/siamese_rpn_r50_20e_lasot.py',
```
Comment (GT9505) [inline tests/test_models/test_forward/test_sot_forward.py:12]: follow [this](https://github.com/open-mmlab/mmtracking/blob/master/tests/test_models/test_track_heads/test_siamese_rpn_head.py) to add ut for stark_hea

**Review by GT9505 — APPROVED** [review]
